core/src/main/java/com/datastrato/gravitino/lock/LockManager.java

```diff
@@ -0,0 +1,200 @@
+              maxNodesInMemory, minNodesInMemory));
+    }
+    this.maxTreeNodeInMemory = maxNodesInMemory;
+    this.minTreeNodeInMemory = minNodesInMemory;
+
+    long cleanIntervalInSecs = config.get(TREE_LOCK_CLEAN_INTERVAL);
+    if (cleanIntervalInSecs <= 0) {
+      throw new IllegalArgumentException(
+          String.format(
+              "The interval in seconds to clean up the stale tree lock nodes '%d' should be greater than 0",
+              cleanIntervalInSecs));
+    }
+
+    this.cleanTreeNodeIntervalInSecs = cleanIntervalInSecs;
+  }
+
+  private void startNodeCleaner() {
+    ScheduledThreadPoolExecutor lockCleaner =
+        new ScheduledThreadPoolExecutor(
+            1,
+            new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat("tree-lock-cleaner-%d")
+                .build());
+
+    lockCleaner.scheduleAtFixedRate(
+        () -> {
+          long nodeCount = totalNodeCount.get();
+          LOG.trace("Total tree lock node count: {}", nodeCount);
+          if (nodeCount > maxTreeNodeInMemory) {
+            treeLockRootNode
+                .getAllChildren()
+                .forEach(child -> evictStaleNodes(child, treeLockRootNode));
+          }
+        },
+        cleanTreeNodeIntervalInSecs,
+        cleanTreeNodeIntervalInSecs,
+        TimeUnit.SECONDS);
+  }
+
+  public LockManager(Config config) {
+    treeLockRootNode = new TreeLockNode(ROOT.name(), this);
+
+    // Init the parameters.
+    initParameters(config);
+
+    // Start tree lock cleaner.
+    startNodeCleaner();
+  }
+
+  /**
+   * Evict the stale nodes from the tree lock node.
+   *
+   * @param treeNode The tree lock node to evict.
+   * @param parent The parent of the tree lock node.
+   */
+  @VisibleForTesting
+  void evictStaleNodes(TreeLockNode treeNode, TreeLockNode parent) {
+    // We will not evict the node tree if the total node count is less than the
+    // MIN_TREE_NODE_IN_MEMORY.
+    // Do not need to consider thread-safe issues.
+    if (totalNodeCount.get() < minTreeNodeInMemory) {
+      return;
+    }
+
+    // Handle from leaf nodes first.
+    treeNode.getAllChildren().forEach(child -> evictStaleNodes(child, treeNode));
+
+    // Handle self node.
+    if (treeNode.getReference() == 0) {
+      synchronized (parent) {
+        // Once goes here, the parent node has been locked, so the reference of child (treeNode)
+        // could not be changed.
+        if (treeNode.getReference() == 0) {
+          parent.removeChild(treeNode.getName());
+          long leftNodeCount = totalNodeCount.decrementAndGet();
+          LOG.trace(
+              "Evict stale tree lock node '{}', current left nodes '{}'",
+              treeNode.getName(),
+              leftNodeCount);
+        } else {
+          treeNode.getAllChildren().forEach(child -> evictStaleNodes(child, treeNode));
+        }
+      }
+    }
```

Comment:
Need add some log for cleanup. total locks,  cleaned locks, time spend.

Reply:
added